// Copyright (c) 2018-2020 MobileCoin Inc.

// MUST BE KEPT IN SYNC WITH RUST CODE!

// mobilecoind client data types and service descriptors.

syntax = "proto3";
import "google/protobuf/empty.proto";
import "external.proto";
import "blockchain.proto";

package mobilecoind_api;

option java_package = "com.mobilecoin.mobilecoind";
option java_outer_classname = "MobileCoinDAPI";

service MobilecoindAPI {
    // Monitors
    rpc AddMonitor (AddMonitorRequest) returns (AddMonitorResponse) {}
    rpc RemoveMonitor (RemoveMonitorRequest) returns (google.protobuf.Empty) {}
    rpc GetMonitorList (google.protobuf.Empty) returns (GetMonitorListResponse) {}
    rpc GetMonitorStatus (GetMonitorStatusRequest) returns (GetMonitorStatusResponse) {}
    rpc GetUnspentTxOutList (GetUnspentTxOutListRequest) returns (GetUnspentTxOutListResponse) {}

    // Utilities
    rpc GenerateEntropy (google.protobuf.Empty) returns (GenerateEntropyResponse) {}
    rpc GetAccountKey (GetAccountKeyRequest) returns (GetAccountKeyResponse) {}
    rpc GetPublicAddress (GetPublicAddressRequest) returns (GetPublicAddressResponse) {}

    // b58 Codes
    rpc ParseRequestCode (ParseRequestCodeRequest) returns (ParseRequestCodeResponse) {}
    rpc CreateRequestCode (CreateRequestCodeRequest) returns (CreateRequestCodeResponse) {}
    rpc ParseTransferCode (ParseTransferCodeRequest) returns (ParseTransferCodeResponse) {}
    rpc CreateTransferCode (CreateTransferCodeRequest) returns (CreateTransferCodeResponse) {}
    rpc ParseAddressCode (ParseAddressCodeRequest) returns (ParseAddressCodeResponse) {}
    rpc CreateAddressCode (CreateAddressCodeRequest) returns (CreateAddressCodeResponse) {}

    // Txs
    rpc GenerateTx (GenerateTxRequest) returns (GenerateTxResponse) {}
    rpc GenerateOptimizationTx (GenerateOptimizationTxRequest) returns (GenerateOptimizationTxResponse) {}
    rpc GenerateTransferCodeTx (GenerateTransferCodeTxRequest) returns (GenerateTransferCodeTxResponse) {}
    rpc GenerateTxFromTxOutList (GenerateTxFromTxOutListRequest) returns (GenerateTxFromTxOutListResponse) {}
    rpc SubmitTx (SubmitTxRequest) returns (SubmitTxResponse) {}

    // Databases
    rpc GetLedgerInfo (google.protobuf.Empty) returns (GetLedgerInfoResponse) {}
    rpc GetBlockInfo (GetBlockInfoRequest) returns (GetBlockInfoResponse) {}
    rpc GetBlock (GetBlockRequest) returns (GetBlockResponse) {}
    rpc GetTxStatusAsSender (GetTxStatusAsSenderRequest) returns (GetTxStatusAsSenderResponse) {}
    rpc GetTxStatusAsReceiver (GetTxStatusAsReceiverRequest) returns (GetTxStatusAsReceiverResponse) {}
    rpc GetProcessedBlock (GetProcessedBlockRequest) returns (GetProcessedBlockResponse) {}
    rpc GetBlockIndexByTxPubKey (GetBlockIndexByTxPubKeyRequest) returns (GetBlockIndexByTxPubKeyResponse) {}

    // Convenience calls
    rpc GetBalance (GetBalanceRequest) returns (GetBalanceResponse) {}
    rpc SendPayment (SendPaymentRequest) returns (SendPaymentResponse) {}
    rpc PayAddressCode (PayAddressCodeRequest) returns (SendPaymentResponse) {}

    // Network status
    rpc GetNetworkStatus (google.protobuf.Empty) returns (GetNetworkStatusResponse) {}
}

//*********************************
//*
//*  Structures
//*
//*********************************


// Possible transaction status values. Senders check with key images. Receivers check with tx public keys.
enum TxStatus {
    // The transaction is not in the public ledger.
    Unknown = 0;

    // The transaction is in the public ledger.
    Verified = 1;

    // Error: The transaction is not in the public ledger, and the tombstone block has been exceeded.
    TombstoneBlockExceeded = 2;

    // The transaction was found in the ledger but the confirmation number is incorrect
    InvalidConfirmationNumber = 3;
}

// Structure used in specifying the list of outputs when generating a transaction.
message Outlay {
    uint64 value = 1;
    external.PublicAddress public_address = 2;
}

// Structure used to refer to a TxOut in the ledger that is presumed to be spendable.
// The structure is annotated with extra information needed to spend the TxOut in a payment, calculated using the private keys that control the TxOut.
message UnspentTxOut {
    // The actual TxOut object found in the ledger.
    external.TxOut tx_out = 1;

    // The subaddress index the TxOut was sent to.
    uint64 subaddress_index = 2;

    // The key image of the TxOut.
    external.KeyImage key_image = 3;

    // The value of the TxOut.
    uint64 value = 4;

    // The ledger block count when this UnspentTxOut was last attempted to be spent.
    uint64 attempted_spend_block_count = 5;

    // The tombstone block used when we attempted to spend the UTXO.
    uint64 attempted_spend_tombstone = 6;

    // The monitor id this UnspentTxOut belongs to.
    // Note that this field is not included in the Rust `utxo_store::UnspentTxOut` struct.
    bytes monitor_id = 10;
}

// Structure used to refer to a prepared transaction
message TxProposal {
    // List of inputs being spent.
    repeated UnspentTxOut input_list = 1;

    // List of outputs being created.
    // This excludes the fee output.
    repeated Outlay outlay_list = 2;

    // The actual transaction object.
    // Together with the private view/spend keys, this structure contains all information in existence about the transaction.
    external.Tx tx = 3;

    // The transaction fee. This is equal to `tx.prefix.fee`.
    // Note that changing this fee will have no effect on the transaction. Changing the fee
    // inside `tx` will invalidate the ring signature.
    uint64 fee = 4;

    /// A map of outlay index -> TxOut index in the Tx object.
    /// This is needed to map receivers to their respective TxOuts.
    map<uint64, uint64> outlay_index_to_tx_out_index = 5;

    /// A list of the confirmation numbers, in the same order
    /// as the outlays.
    repeated bytes outlay_confirmation_numbers = 6;
}

// Structure used to check transaction status as a Sender.
message SenderTxReceipt {
    // Key images that are going to be added to the ledger once the transaction goes through.
    repeated external.KeyImage key_image_list = 1;

    // Tombstone block set in the transaction.
    uint64 tombstone = 2;
}

// Structure used to check transaction status as a receiver.
// There exists one receipt per output, so a transaction having multiple outputs would have
// multiple ReceiverTxReceipts.
message ReceiverTxReceipt {
    // The receiver this receipt refers to
    external.PublicAddress public_address = 1;

    // The public key of the TxOut sent to this receiver.
    external.CompressedRistretto tx_out_public_key = 2;

    // The hash of the TxOut sent to this receiver.
    bytes tx_out_hash = 3;

    // Tombstone block set in the transaction.
    uint64 tombstone = 4;

    // Confirmation number for this TxOut
    bytes confirmation_number = 5;
}

// Structure used to report monitor status
message MonitorStatus {
    // The account key the monitor is monitoring.
    external.AccountKey account_key = 1;

    // The first subaddress index being monitored.
    uint64 first_subaddress_index = 2;

    // The number of subaddresses being monitored, starting at first_subaddress.
    uint64 num_subaddresses = 3;

    // Block index we started scanning from.
    uint64 first_block_index = 4;

    // Next block index we are waiting to sync.
    uint64 next_block_index = 5;

    // Optional monitor name.
    string name = 6;
}

// Enum used to indicate whether a ProcessedTxOut is a sent one or a received one.
enum ProcessedTxOutDirection {
    // This should never happen, but is available here as an option to catch uninitialized data.
    // The name "Unknown" cannot be used because, quoting the protobuf compiler:
    // Note that enum values use C++ scoping rules, meaning that enum values are siblings of their type, not children of it.  Therefore, "Unknown" must be unique within "mobilecoind_api", not just within "ProcessedTxOutDirection".
    Invalid = 0;

    // The ProcessedTxOut has been received at the block queried for.
    Received = 1;

    // The ProcessedTxOut has been spent at the block queried for.
    Spent = 2;
}


// Structure used to report processed information for TxOuts discovered in a given processed block.
message ProcessedTxOut {
    // The monitor id that owns the TxOut.
    bytes monitor_id = 1;

    // The subaddress that owns the TxOut.
    uint64 subaddress_index = 2;

    // The public key of the TxOut.
    external.CompressedRistretto tx_out_public_key = 3;

    // The key image of the TxOut.
    external.KeyImage key_image = 4;

    // The value of the TxOut.
    uint64 value = 5;

    // Whether the TxOut was received (deposit to subaddress) or spent (withdrawal from subaddress).
    ProcessedTxOutDirection direction = 6;

    // The b58-encoded Address Code for the subaddress that owns the TxOut.
    string b58_code = 7;
}

//*********************************
//*
//*  Requests and Responses for API
//*
//*********************************

//
// Monitors
//

// Add a new monitor.
message AddMonitorRequest {
    // Account key to monitor.
    external.AccountKey account_key = 1;

    // The first subaddress index being monitored.
    uint64 first_subaddress_index = 2;

    // The number of subaddresses being monitored, starting at first_subaddress.
    uint64 num_subaddresses = 3;

    // Block index to start monitoring from.
    uint64 first_block_index = 4;

    // Optional name.
    string name = 5;
}
message AddMonitorResponse {
    bytes monitor_id = 1;
    bool is_new = 2;
}

// Remove a monitor and all associated data.
message RemoveMonitorRequest {
    bytes monitor_id = 1;
}
// - empty response

// List all known monitor ids.
// - empty request
message GetMonitorListResponse {
    repeated bytes monitor_id_list = 1;
}

// Get the status of a specific monitor.
message GetMonitorStatusRequest {
    bytes monitor_id = 1;
}
message GetMonitorStatusResponse {
    MonitorStatus status = 1;
}

// Get a list of UnspentTxOuts for a given monitor and subadddress index.
message GetUnspentTxOutListRequest {
    bytes monitor_id = 1;
    uint64 subaddress_index = 2;
}
message GetUnspentTxOutListResponse {
    repeated UnspentTxOut output_list = 1;
}

//
// Utilities
//

// Generate a new random root entropy value.
// - empty request
message GenerateEntropyResponse {
	  // 32 bytes generated using a cryptographically secure RNG.
    bytes entropy = 1;
}

// Generate an AccountKey from a 32 byte root entropy value.
message GetAccountKeyRequest {
    bytes entropy = 1;
}
message GetAccountKeyResponse {
    external.AccountKey account_key = 1;
}

// Get the public address for a given monitor and subadddress index.
message GetPublicAddressRequest {
    bytes monitor_id = 1;
    uint64 subaddress_index = 2;
}
message GetPublicAddressResponse {
    external.PublicAddress public_address = 1;
    string b58_code = 2;
}

//
// b58 Codes
//

// Decode a base-58 encoded "MobileCoin Request Code" into receiver's public address, value, and memo.
message ParseRequestCodeRequest {
    string b58_code = 1;
}
message ParseRequestCodeResponse {
    external.PublicAddress public_address = 1;
    uint64 value = 2;
    string memo = 3;
}

// Encode receiver's public address, value, and memo into a base-58 "MobileCoin Request Code".
message CreateRequestCodeRequest {
    external.PublicAddress public_address = 1;
    uint64 value = 2;
    string memo = 3;
}
message CreateRequestCodeResponse {
    string b58_code = 1;
}

// Decode a base-58 encoded "MobileCoin Transfer Code" into entropy/tx_out_public_key/memo.
// This code provides a mobile client with everything required to construct a self-payment, allowing funds to be withdrawn from a gift card.
message ParseTransferCodeRequest {
    string b58_code = 1;
}
message ParseTransferCodeResponse {
    bytes entropy = 1;
    external.CompressedRistretto tx_out_public_key = 2;
    string memo = 3;
    UnspentTxOut utxo = 4;
}

// Encode entropy/tx_out_public_key/memo into a base-58 "MobileCoin Transfer Code".
message CreateTransferCodeRequest {
    bytes entropy = 1;
    external.CompressedRistretto tx_out_public_key = 2;
    string memo = 3;
}
message CreateTransferCodeResponse {
    string b58_code = 1;
}

// Decode a base-58 encoded "MobileCoin Address Code" into the receiver's public address.
message ParseAddressCodeRequest {
    string b58_code = 1;
}
message ParseAddressCodeResponse {
    external.PublicAddress public_address = 1;
}

// Encode receiver's public address into a base-58 "MobileCoin Address Code".
message CreateAddressCodeRequest {
    external.PublicAddress public_address = 1;
}
message CreateAddressCodeResponse {
    string b58_code = 1;
}

//
// Transactions
//

// Generate a transaction proposal object.
// Notes:
// - Sum of inputs needs to be greater than sum of outlays and fee.
// - The set of inputs to use would be chosen automatically by mobilecoind.
// - The fee field could be set to zero, in which case mobilecoind would choose a fee.
// Right now that fee is hardcoded.
message GenerateTxRequest {
    // Monitor id sending the funds.
    bytes sender_monitor_id = 1;

    // Subaddress index to return change to.
    uint64 change_subaddress_index = 2;

    // List of UnspentTxOuts to be spent by the transaction.
    // All UnspentTxOuts must belong to the same sender_monitor_id.
    // mobilecoind would choose a subset of these inputs to construct the transaction.
    // Total input amount must be >= sum of outlays + fees.
    repeated UnspentTxOut input_list = 3;

    // Outputs to be generated by the transaction. This excludes change and fee.
    repeated Outlay outlay_list = 4;

    // Fee in picoMOB (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in tx_proposal.tx.prefix.fee
    uint64 fee = 5;

    // Tombstone block (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in tx_proposal.tx.prefix.tombstone_block
    uint64 tombstone = 6;
}
message GenerateTxResponse {
    TxProposal tx_proposal = 1;
}

// Generate a transaction that merges a few UnspentTxOuts into one, in order to reduce wallet fragmentation.
message GenerateOptimizationTxRequest {
    // Monitor Id to operate on.
    bytes monitor_id = 1;

    // Subaddress to operate on.
    uint64 subaddress_index = 2;
}
message GenerateOptimizationTxResponse {
    TxProposal tx_proposal = 1;
}

// Generate a transaction that can be used to fund a "MobileCoin Transfer Code"
message GenerateTransferCodeTxRequest {
    // Monitor id sending the funds.
    bytes sender_monitor_id = 1;

    // Subaddress index to return change to.
    uint64 change_subaddress_index = 2;

    // List of UnspentTxOuts available for this transaction. All must belong to sender_monitor_id.
    // mobilecoind will choose a subset of these inputs to construct the transaction.
    // Total value must be >= sum of outlays + fees.
    repeated UnspentTxOut input_list = 3;

    // Value to move to the "Transfer Code" in picoMOB
    uint64 value = 4;

    // Fee in picoMOB (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in the resulting tx_proposal.tx.prefix.fee
    uint64 fee = 5;

    // Tombstone block (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in tx_proposal.tx.prefix.tombstone_block
    uint64 tombstone = 6;

    // The memo to be included in the Transfer Code
    string memo = 7;
}
message GenerateTransferCodeTxResponse {
    // The tx proposal to submit to the network.
    TxProposal tx_proposal = 1;

    // The entropy for constructing the AccountKey that can access the funds.
    bytes entropy = 2;

    // The TxOut public key that has the funds.
    external.CompressedRistretto tx_out_public_key = 3;

    // The memo (simply copied from the request).
    string memo = 4;

    // The b58-encoded Transfer Code
    string b58_code = 5;
}

// Generate a transaction without a monitor, requires an account key and
// a list of UnspentTxOuts. All coins (minus the fee) are transferred to
// a single recipient. Used for temporary accounts like gift codes.
message GenerateTxFromTxOutListRequest {
    // Account key that owns the transactions
    external.AccountKey account_key = 1;

    // List of TxOuts to spend
    repeated UnspentTxOut input_list = 2;

    // Address to transfer coins to
    external.PublicAddress public_address = 3;

    // Fee
    uint64 fee = 4;
}
message GenerateTxFromTxOutListResponse {
    TxProposal tx_proposal = 1;
}

// Submits a generated transaction proposal to the network.
message SubmitTxRequest {
    TxProposal tx_proposal = 1;
}
message SubmitTxResponse {
    SenderTxReceipt sender_tx_receipt = 1;
    repeated ReceiverTxReceipt receiver_tx_receipt_list = 2;
}

//
// Databases
//

// Get information about the downloaded ledger.
// - empty request
message GetLedgerInfoResponse {
    // Total number of blocks in the ledger.
    uint64 block_count = 1;

    // Total number of TxOuts in the ledger.
    uint64 txo_count = 2;
}

// Get information about a downloaded block.
message GetBlockInfoRequest {
    // the block index of interest
    uint64 block_index = 1;
}
message GetBlockInfoResponse {
    // Number of key images in the block.
    uint64 key_image_count = 1;

    // Number of TxOuts in the block.
    uint64 txo_count = 2;
}

message ArchiveBlockSignatureData {
    // The origin of the Archive Block.
    string src_url = 1;

    // The archive filename.
    string filename = 2;

    // The block signature.
    blockchain.BlockSignature signature = 3;
}

// Get more detailed information about a downloaded block
message GetBlockRequest {
    // the block index of interest
    uint64 block_index = 1;
}
message GetBlockResponse {
    // The block
    blockchain.Block block = 1;

    // Signatures for this block
    repeated ArchiveBlockSignatureData signatures = 2;

    // Key images in the block
    repeated external.KeyImage key_images = 3;

    // TxOuts in the block.
    repeated external.TxOut txos = 4;
}

// Get the status of a submitted transaction as the Sender (using the key image).
message GetTxStatusAsSenderRequest {
    SenderTxReceipt receipt = 1;
}
message GetTxStatusAsSenderResponse {
    TxStatus status = 1;
}

// Get the status of a submitted transaction as the receiver (using the tx public key).
message GetTxStatusAsReceiverRequest {
    ReceiverTxReceipt receipt = 1;

    // Optionally pass in a monitor ID to validate confirmation number
    bytes monitor_id = 2;
}
message GetTxStatusAsReceiverResponse {
    TxStatus status = 1;
}

// Get the contents of a processed block.
message GetProcessedBlockRequest {
    // Monitor id to query data for.
    bytes monitor_id = 1;

    // Block index to query.
    uint64 block_index = 2;
}
message GetProcessedBlockResponse {
    // Processed tx output information that belongs to the requested monitor_id/block.
    repeated ProcessedTxOut tx_outs = 1;
}

// Get the block index containing a given TxOut public key.
message GetBlockIndexByTxPubKeyRequest {
    // The TxOut public key to look for.
    external.CompressedRistretto tx_out_public_key = 1;
}
message GetBlockIndexByTxPubKeyResponse {
    // The block index.
    uint64 block_index = 1;
}

//
// Convenience calls
///

// Get the balance for a given monitor and subadddress index, in picoMOB.
message GetBalanceRequest {
    // Monitor id to query balance for.
    bytes monitor_id = 1;

    // Subaddress to query balance for.
    uint64 subaddress_index = 2;
}
message GetBalanceResponse {
    // Sum of all utxos associated with the requested monitor_id/subaddress_index.
    uint64 balance = 1;
}

// Build and submit a simple payment and return any change to the Sender's subaddress.
message SendPaymentRequest {
    // Monitor id sending the funds.
    bytes sender_monitor_id = 1;

    // Subaddress the funds are coming from.
    uint64 sender_subaddress_index = 2;

    // Outputs to be generated by the transaction. This excludes change and fee.
    repeated Outlay outlay_list = 3;

    // Fee in picoMOB (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in tx_proposal.tx.prefix.fee
    uint64 fee = 4;

    // Tombstone block (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in tx_proposal.tx.prefix.tombstone_block
    uint64 tombstone = 5;

    // Optional: When selecting input UTXOs for the transaction, limit selection only to UTXOs whose
    // value is lower or equal to to this.
    uint64 max_input_utxo_value = 6;
}
message SendPaymentResponse {
    // Information the sender can use to check if the transaction landed in the ledger.
    SenderTxReceipt sender_tx_receipt = 1;

    // Information receivers can use to check if the transaction landed in the ledger.
    repeated ReceiverTxReceipt receiver_tx_receipt_list = 2;

    // The TxProposal that was submitted to the network. The fee that was paid can be checked at
    // tx_proposal.tx.prefix.fee
    TxProposal tx_proposal = 3;
}

// Build and submit a simple payment to an address provided by a b58 address code
message PayAddressCodeRequest {
    // Monitor id sending the funds.
    bytes sender_monitor_id = 1;

    // Subaddress the funds are coming from.
    uint64 sender_subaddress_index = 2;

    // Base-58 encoded "MobileCoin Address Code"
    string b58_code = 3;

    // Amount to pay
    uint64 amount = 4;

    // Fee in picoMOB (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in tx_proposal.tx.prefix.fee
    uint64 fee = 5;

    // Tombstone block (setting to 0 causes mobilecoind to choose a value).
    // The value used can be checked (but not changed) in tx_proposal.tx.prefix.tombstone_block
    uint64 tombstone = 6;

    // Optional: When selecting input UTXOs for the transaction, limit selection only to UTXOs whose
    // value is lower or equal to to this.
    uint64 max_input_utxo_value = 7;
}

//
// Network status
//

// Get information about the network.
// - empty request
message GetNetworkStatusResponse {
    // Total highest block number the network agrees on.
    // (This is the block number we will try to sync to).
    uint64 network_highest_block_index = 1;

    // A map of node responder id to the block index reported by it.
    map<string, uint64> peer_block_index_map = 2;

    // The local ledger block index.
    uint64 local_block_index = 3;

    // Whether we are behind.
    bool is_behind = 4;
}
